File: 733.flood_fill.java
class Solution() {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if (image[sr][sc] != newColor)
            dfs(image, sr, sc, image[sr][sc], newColor);
        return image;
    }
    public void dfs(int[][] grid, int x, int y, int initial, int newColor) {
        if (image[x][y] != initial)
            return;
        image[r][c] = newColor;
        if (x > 0) dfs(grid, x - 1, y, intial, newColor);
        if (y > 0) dfs(grid, x, y - 1, intial, newColor);
        if (x + 1 < grid.length()) dfs(grid, x + 1, y, initial, newColor);
        if (y + 1 < grid[0].length()) dfs(grid, x, y + 1, initial, newColor);
    }
}

File: pres/733.flood_fill.cpp
#include <vector>
#include <stdlib.h>
using namespace std;



class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        if (image[sr][sc] == color) return image;

        dfs(image, sr, sc, image[sr][sc], color);

        return image;
    }

    void dfs (vector<vector<int>>& grid, int x, int y, int pre_replace, int post_replace) {
        if (grid[x][y] != pre_replace || x < 0 || y < 0 || x >= grid.size() || y >= grid[0].size())
            return;
        grid[x][y] = post_replace;

        for (int x_off = -1; x_off <= 1; x_off++)
            for (int y_off = -1; y_off <= 1; y_off++)
                if (abs(x_off) != abs(y_off))
                    dfs(grid, x + x_off, y_off, pre_replace, post_replace);
    }
};
